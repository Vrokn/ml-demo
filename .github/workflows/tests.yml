name: tests

on: push

jobs:
  unit:
    runs-on: ubuntu-latest

    steps:
      - uses: actions/checkout@v3

      - uses: actions/setup-node@v3
        with:
          node-version: 16
          cache: "yarn"

      - name: Install Dependencies
        run: yarn

      - name: Test
        run: npx jest --config=jest.config.js
