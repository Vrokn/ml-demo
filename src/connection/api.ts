/* eslint-disable @typescript-eslint/no-explicit-any */
import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';

export const BASE_URL = 'https://music.musicaudience.info';
export const API_KEY = '5db48e1f3a0a4580bad47849f1317bd0';

type ApiRequestConfig = AxiosRequestConfig;

export const apiCall = async (
  config: ApiRequestConfig,
  showSnackbar?: (text: string, type: "success" | "info" | "warning" | "error") => void
): Promise<AxiosResponse<any> | null> => {
  const headers = {
    Authorization: `apikey ${API_KEY}`,
    ...config.headers,
  };

  try {
    return await axios({
      ...config,
      baseURL: BASE_URL,
      headers,
    });
  } catch (error: any) {
    console.error("API call error:", error);

    let errorMessage = 'Oops! Something went wrong with the request. Please try again.';
    if (error.response && error.response.data) {
      errorMessage = error.response.data.error || errorMessage;
    }
    showSnackbar && showSnackbar(errorMessage, 'error');
    return null;
  }
};