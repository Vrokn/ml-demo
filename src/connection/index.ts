export * from './genres';
export * from './artists';
