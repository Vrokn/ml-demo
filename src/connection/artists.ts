import { apiCall } from './api';

export const searchArtists = (query: string, limit?: number) => {
  return apiCall({
    method: 'GET',
    url: '/api/v1/music/artists',
    params: { q: query, limit },
  });
};

export const getArtistDetails = (id: number) => {
  return apiCall({
    method: 'GET',
    url: `/api/v1/music/artists/${id}`,
  });
};

export const getSimilarArtists = (id: number) => {
  return apiCall({
    method: 'GET',
    url: `/api/v1/music/artists/${id}/similar`,
  });
};
// Other artist-related functions
