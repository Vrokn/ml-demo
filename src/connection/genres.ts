import { apiCall } from './api';

export const getGenres = (query: string, limit?: number) => {
  return apiCall({
    method: 'GET',
    url: '/api/v1/music/genres',
    params: { q: query, limit },
  });
};

export const getArtistByGenre = (id: number) => {
  return apiCall({
    method: 'GET',
    url: `/api/v1/music/genres/${id}/artists`,
  });
};
