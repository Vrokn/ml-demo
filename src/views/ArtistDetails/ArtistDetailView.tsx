import React, { useEffect, useMemo, useState } from 'react';
import { useParams } from 'react-router-dom';
import { Box, Typography, Chip, Link, Grid, LinearProgress, Button } from '@mui/material';
import { getArtistDetails, getSimilarArtists } from '../../connection';
import { useSnackbar } from '../../context/SnackbarContext/SnackbarContext';
import { useArtistContext } from '../../context/SavedArtistContext/SavedArtistContext';
import StarBorderIcon from '@mui/icons-material/StarBorder';
import StarIcon from '@mui/icons-material/Star';
import { Artist } from '../../interfaces/artist';
import GenreChips from '../../components/GenreChips/GenreChips';
import { Genre } from '../../interfaces/genre';
import { ChipsContainer } from '../../components/GenreChips/GenreChips.styles';

const ArtistDetailView = () => {
  const { artistId } = useParams();
  const [artist, setArtist] = useState<Artist | null>(null);
  const [similarArtists, setSimilarArtists] = useState([]);
  const { artists, toggleArtist } = useArtistContext();

  const { showSnackbar } = useSnackbar();

  useEffect(() => {
    if (artistId) {
      getArtistDetails(parseInt(artistId))
        .then((response) => {
          setArtist(response?.data.data?.[0]);
        })
        .catch((error) => {
          showSnackbar(error?.message, 'error');
          console.error('An error occurred while fetching artist details:', error);
        });
      getSimilarArtists(parseInt(artistId))
        .then((response) => {
          setSimilarArtists(response?.data.data);
        })
        .catch((error) => {
          showSnackbar(error?.message, 'error');
          console.error('An error occurred while fetching related artist details:', error);
        });
    }
  }, [artistId]);

  const primaryGenres = useMemo(() => {
    return artist?.genres.filter(item => item.is_primary) || [];
  }, [artist]);

  const secondaryGenres = useMemo(() => {
    return artist?.genres.filter(item => item.is_primary !== 1) || [];
  }, [artist]);

  const isPresent = (id: number) => artists.some(a => a.id === id);

  const handleClick = (artist: Artist) => {
    toggleArtist(artist);
  };

  return (
    <Box display="flex" flex={1} justifyContent="space-between">
      {(!artist || !similarArtists) && (<LinearProgress />)}
      <Box sx={{ padding: '2em', display: 'flex', flexDirection: 'column', justifyContent: 'center' }}>
        {artist && (
          <Box display="flex" flexDirection="column" gap="1em"  >
            <Typography
              fontSize="3vw"
              fontWeight="bold"
              maxWidth="40vw"
              sx={{
                whiteSpace: 'nowrap',
                overflow: 'hidden',
                textOverflow: 'ellipsis',
              }}
            >
              {artist.name.toLocaleUpperCase()} :
            </Typography>
            <img src={artist.image} alt={artist.name} style={{
              height: "55vh",
              width: '40vw',
              objectFit: 'cover',
              borderRadius: '4px'
            }}
              onError={(e) => {
                (e.target as HTMLImageElement).src = 'https://placehold.co/400?text=No+image';
              }}
            />
            <Button
              children={isPresent(artist.id) ? "Remove" : "Add"}
              onClick={() => handleClick(artist)}
              startIcon={isPresent(artist.id) ? <StarIcon /> : <StarBorderIcon />}
              variant={isPresent(artist.id) ? "contained" : "outlined"}
            />
            {primaryGenres?.length > 0 && (
              <GenreChips
                genres={primaryGenres}
                label={"Primary genres"}
                color={"primary"}
              />
            )}
            {secondaryGenres?.length > 0 && (
              <GenreChips
                genres={secondaryGenres}
                label={"Additional genres"}
                color={"secondary"}
              />
            )}
            <ChipsContainer>
              <Typography variant="caption">Popularity score:</Typography>
              <Typography variant="h6" fontWeight="bold">{artist.popularity}</Typography>
            </ChipsContainer>
          </Box>
        )}
      </Box>
      {
        similarArtists.length > 0 && (<Box sx={{ overflow: 'auto', padding: '2em', maxHeight: '85vh', width: '60%' }} my={0.5}>
          <Typography variant='h5' color="secondary">Similar artists:</Typography>
          <Grid container>
            {similarArtists.map((similarArtist: Artist) => (
              <Grid item xl={6} sm={12} key={similarArtist.id}>
                <Box display="flex" gap="2em" alignItems="center" width="100%" my={1}>
                  <Box display="flex" flexDirection="column">
                    <img src={similarArtist.image} alt={similarArtist.name} style={{
                      height: '140px',
                      width: '140px',
                      objectFit: 'cover',
                      borderRadius: '4px'
                    }}
                      onError={(e) => {
                        (e.target as HTMLImageElement).src = 'https://placehold.co/400?text=No+image';
                      }}
                    />
                    <Button
                      fullWidth
                      children={isPresent(similarArtist.id) ? "Remove" : "Add"}
                      onClick={() => handleClick(similarArtist)}
                      startIcon={isPresent(similarArtist.id) ? <StarIcon /> : <StarBorderIcon />}
                      variant={isPresent(similarArtist.id) ? "contained" : "outlined"}
                      size='small'
                    />
                  </Box>
                  <Box display="flex" flexDirection="column">
                    <Typography variant="h6"><Link href={`/artist/${similarArtist.id}`}>{similarArtist.name}</Link></Typography>
                    <Typography variant="body2" color="secondary" >
                      Genres: {similarArtist.genres.map((genre: Genre) =>
                        <Chip
                          key={genre.id}
                          label={genre.name}
                          color='primary'
                          size="small"
                          variant='outlined'
                        />)}
                    </Typography>
                  </Box>
                </Box>
              </Grid>
            ))}
          </Grid>
        </Box>
        )}
    </Box >
  );
};

export default ArtistDetailView;
