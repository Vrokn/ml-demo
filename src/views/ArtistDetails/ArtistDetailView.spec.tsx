import React from 'react';
import { screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { getArtistDetails, getSimilarArtists } from '../../connection';
import ArtistDetailView from './ArtistDetailView';
import { render } from '../../../tests/testUtils';

const artistData = { id: 1, name: 'Artist Name', image: 'image-url', popularity: 50, genres: [] };
const similarArtistsData = [{ id: 2, name: 'Similar Artist', image: 'image-url', genres: [] }];

jest.mock('../../connection', () => ({
  getArtistDetails: jest.fn(() => Promise.resolve({ data: { data: [artistData] } })),
  getSimilarArtists: jest.fn(() => Promise.resolve({ data: { data: similarArtistsData } })),
}));


describe('ArtistDetailView', () => {

  it('renders artist details and matches snapshot', () => {
    const { asFragment } = render(<ArtistDetailView />);
    expect(asFragment()).toMatchSnapshot();
  });

 /*  it('displays artist details', async () => {
    (getArtistDetails as jest.Mock).mockResolvedValue({ data: { data: [artistData] } });
    (getSimilarArtists as jest.Mock).mockResolvedValue({ data: { data: similarArtistsData } });

    render(<ArtistDetailView />);

    const artistNameElement = await screen.findByText(artistData.name.toLocaleUpperCase());
    expect(artistNameElement).toBeInTheDocument();
  });

  it('displays similar artists', async () => {
    (getArtistDetails as jest.Mock).mockResolvedValue({ data: { data: [artistData] } });
    (getSimilarArtists as jest.Mock).mockResolvedValue({ data: { data: similarArtistsData } });

    render(<ArtistDetailView />);
    const similarElement = await screen.findByText('Similar artists:');
    expect(similarElement).toBeInTheDocument();
  }); */
});
