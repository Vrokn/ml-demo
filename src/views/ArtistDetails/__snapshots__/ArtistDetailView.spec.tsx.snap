// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistDetailView renders artist details and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-zk51sn"
  >
    <span
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-ug1ls-MuiLinearProgress-root"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-1f2ktpx-MuiLinearProgress-bar1"
      />
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1rvn8xp-MuiLinearProgress-bar2"
      />
    </span>
    <div
      class="MuiBox-root css-3kv8uz"
    />
  </div>
</DocumentFragment>
`;
